hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/DeltaSync.java

```diff
@@ -698,8 +717,13 @@ private Pair<Option<String>, JavaRDD<WriteStatus>> writeToSink(JavaRDD<HoodieRec
     boolean hasErrors = totalErrorRecords > 0;
     if (!hasErrors || cfg.commitOnErrors) {
       HashMap<String, String> checkpointCommitMetadata = new HashMap<>();
+      Option<BiConsumer<HoodieTableMetaClient, HoodieCommitMetadata>> extraPreCommitFunc = Option.empty();
       if (checkpointStr != null) {
-        checkpointCommitMetadata.put(CHECKPOINT_KEY, checkpointStr);
+        if (identifier.isPresent()) {
+          extraPreCommitFunc = Option.of(new HoodieDeltaStreamerMultiwriterCheckpoint(this, checkpointStr, latestCheckpointWritten));
```

Comment:
fetching of latestCheckpoint should happen within the extraPreCommitFunc i.e. within the lock.

Reply:
That does happen inside. latestCheckpointWritten is used to detect if multiple deltastreamers are using the same id. We keep track of the last checkpoint that we wrote so that if it doesn't match the latest commit, then we know that another deltastreamer used the same id